src/subscriber.ts

```diff
@@ -170,52 +183,113 @@ export class Subscriber {
 
     private async _sessionOfflineHandler(header: Header): Promise<void> {
       const isHeartbeatExpected = await this._isHeadAfterHeartbeatBlockThreshold(header)
-      const sessionIndex = await this.api.query.session.currentIndex()
+      const sessionIndex = this.sessionIndex 
+
+      this.validators.forEach(async account => {
 
-     this.validators.forEach(async account => {
-        if( isHeartbeatExpected && ! await this._hasValidatorAuthoredBlocks(account,sessionIndex) && ! await this._hasValidatorSentHeartbeat(account,sessionIndex) ){
-          this.logger.info(`Target ${account.name} has either not authored any block or sent any heartbeat yet`);
-          this.promClient.setStateValidatorOfflineSessionReports(account.name)
+        const validatorActiveSetIndex = this._getValidatorActiveSetIndex(account)
```

Comment:
We are going to do this check for each block, but we should only do it at the beginning of the era, i think it's ok for now to avoid extending more this PR, but we could create an issue to address in a follow up.

Reply:
I agree with you. The nice thing is that this check is now completely local and it does not involve any network calls to the node anymore: the needed information are retrieved at each SESSION update via event subscription ( I'd have preferred an ERA based event, but unfortunately it is not provided by the library; still I think it's a nice optimization compared to the previous situation ) wdyt ??

Reply:
Indeed, it is now clear in `_newSessionEventHandler` lgtm!